src/tuf/trustroot.ts

```diff
@@ -0,0 +1,159 @@
+
+const TRUSTED_ROOT_MEDIA_TYPE =
+  'application/vnd.dev.sigstore.trustedroot+json;version=0.1';
+
+// Type describing the Sigstore-specific metadata for a TUF target
+interface SigstoreTargetMetadata {
+  status: string;
+  usage: string;
+  uri: string;
+}
+
+// Type guard for SigstoreTargetMetadata
+function isTargetMetadata(m: unknown): m is SigstoreTargetMetadata {
+  return (
+    m !== undefined &&
+    m !== null &&
+    typeof m === 'object' &&
+    'status' in m &&
+    'usage' in m &&
+    'uri' in m
+  );
+}
+
+export class TrustedRootFetcher {
+  private tuf: Updater;
+  constructor(tuf: Updater) {
+    this.tuf = tuf;
+  }
+
+  // Assembles a TrustedRoot from the targets in the TUF repo
+  async getTrustedRoot(): Promise<sigstore.TrustedRoot> {
+    // Get all available targets
+    const targets = await this.allTargets();
+
+    const cas = await this.getCAKeys(targets, 'Fulcio');
+    const ctlogs = await this.getTLogKeys(targets, 'CTFE');
+    const tlogs = await this.getTLogKeys(targets, 'Rekor');
+
+    return {
+      mediaType: TRUSTED_ROOT_MEDIA_TYPE,
+      certificateAuthorities: cas,
+      ctlogs: ctlogs,
+      tlogs: tlogs,
+      timestampAuthorities: [],
+    };
+  }
+
+  // Retrieves the list of TUF targets.
+  // NOTE: This is a HACK to get around the fact that the TUF library doesn't
+  // expose the list of targets. This is a temporary solution until TUF comes up
+  // with a story for target discovery.
+  // https://docs.google.com/document/d/1rWHAM2qCUtnjWD4lOrGWE2EIDLoA7eSy4-jB66Wgh0o
+  private async allTargets(): Promise<TargetFile[]> {
+    try {
+      await this.tuf.refresh();
+    } catch (e) {
+      throw new TrustedRootError('error refreshing trust metadata');
+    }
+
+    return Object.values(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (this.tuf as any).trustedSet.targets?.signed.targets || {}
+    );
+  }
+
+  // Filters the supplied list of targets to those with the specified usage
+  // and returns a new TransparencyLogInstance for each with the associated
+  // public key populated.
+  private async getTLogKeys(
+    targets: TargetFile[],
+    usage: string
+  ): Promise<sigstore.TransparencyLogInstance[]> {
+    const filteredTargets = filterByUsage(targets, usage);
+
+    return Promise.all(
+      filteredTargets.map(async (target) => {
+        const keyBytes = await this.readKey(target);
+        const uri = isTargetMetadata(target.custom.sigstore)
+          ? target.custom.sigstore.uri
+          : '';
+
+        // The log ID is not present in the Sigstore target metadata, but
+        // can be derived by hashing the contents of the public key.
+        return {
+          baseUrl: uri,
+          hashAlgorithm: sigstore.HashAlgorithm.SHA2_256,
+          logId: { keyId: crypto.hash(keyBytes) },
+
+          publicKey: {
+            keyDetails: sigstore.PublicKeyDetails.PKIX_ECDSA_P256_SHA_256,
+            rawBytes: keyBytes,
```

Comment:
This is not really right? The Key on disk is PEM encoded but in the bundle we only expect the base64 encoded DER bytes, i.e the PEM headers should be stripped.

Reply:
Sorry, missed that `readKey` extracts the correct bytes.